.github/workflows/java-build-for-release.yml

```
@@ -5,10 +5,34 @@ on:
       # if you change this pattern, make sure jobs.strip-tag still works
       - 'release/java/v[0-9]+.[0-9]+.[0-9]+'
 jobs:
+  check-tag-main:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout all branches
+        uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
+        with:
+          fetch-depth: 0
+
+      - name: Verify tag on branch
+        # Check main is one of the branches where the tagged commit exists
+        run: |
+          branches=$(git branch --contains ${{ github.ref }}) --format "%(refname:short)"
+          found=false
+          for b in $raw; do
```

Slight nitpick: you can expand into an array here and then use a bit of dedicated syntax to expand into the loop more idiomatically (IMO):

```suggestion
          branches=($(git branch --contains ${{ github.ref }}) --format "%(refname:short)")
          found=false
          for b in "${branches[@]}"; do
```